Implement Subscription Query Executor

wedamija opened the pull request:
This implements the `Executor` class, which accepts a `ScheduledTask[Subscription]` and returns a
`Future[ClickhouseQueryResult]`. We run the queries in a `ThreadPoolExecutor`, and I've verified
that the code should work fine when run multithreaded.

wedamija committed:
Implement Subscription Query Executor

This implements the `Executor` class, which accepts a `ScheduledTask[Subscription]` and returns a
`Future[ClickhouseQueryResult]`. We run the queries in a `ThreadPoolExecutor`, and I've verified
that the code should work fine when run multithreaded.

wedamija commented:
Hacked together a quick test to run queries in parallel:

```
from unittest.mock import Mock

import calendar
import uuid
from datetime import datetime, timedelta

from snuba import settings
from snuba.datasets.factory import enforce_table_writer
from snuba.subscriptions.consumer import Tick
from snuba.subscriptions.data import Subscription
from snuba.subscriptions.executor import Executor
from snuba.subscriptions.scheduler import ScheduledTask
from snuba.utils.types import Interval
from tests.base import BaseEventsTest


class TestExecutorBulk(BaseEventsTest):
    def setup_method(self, test_method, dataset_name="events"):
        super().setup_method(test_method, dataset_name)
        self.project_id = 1
        self.platforms = ["a", "b"]
        self.minutes = 6000

        self.base_time = datetime.utcnow().replace(
            minute=0, second=0, microsecond=0
        ) - timedelta(minutes=self.minutes)
        self.generate_events()

    def generate_events(self):
        events = []
        for tick in range(self.minutes):
            events.append(
                enforce_table_writer(self.dataset)
                .get_stream_loader()
                .get_processor()
                .process_insert(
                    {
                        "project_id": self.project_id,
                        "event_id": uuid.uuid4().hex,
                        "deleted": 0,
                        "datetime": (self.base_time + timedelta(minutes=tick)).strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
                        "message": "a message",
                        "platform": self.platforms[tick % len(self.platforms)],
                        "primary_hash": uuid.uuid4().hex,
                        "group_id": tick,
                        "retention_days": settings.DEFAULT_RETENTION_DAYS,
                        "data": {
                            "received": calendar.timegm((self.base_time + timedelta(minutes=tick)).timetuple()),
                        },
                    }
                )
            )
        self.write_processed_records(events)

    def test_bulk(self):
        subscription = Subscription(
            id="hello",
            project_id=self.project_id,
            conditions=[],
            aggregations=[["count()", "", "count"]],
            time_window=timedelta(minutes=self.minutes),
            resolution=timedelta(minutes=1),
        )
        task = ScheduledTask(self.base_time, subscription)
        futures = []
        for i in range(self.minutes):
            executor = Executor(self.dataset)
            tick = Tick(
                offsets=Interval(0, 9000),
                timestamps=Interval(self.base_time, self.base_time + timedelta(minutes=i)),
            )

            futures.append(executor.execute(task, tick, Mock()))
            # assert executor.execute(task, tick, Mock()).result ()['data'][0]['count'] == i, i

        for i, future in enumerate(futures):
            result = future.result()
            assert result['data'][0]['count'] == i, (i, result)
```

wedamija committed:
Merge branch 'master' into danf/subscription_executor

tkaemming commented on snuba/subscriptions/executor.py:
```diff
@@ -0,0 +1,38 @@
+from snuba.utils.metrics.timer import Timer
+
+
+executor_pool = ThreadPoolExecutor(max_workers=settings.SUBSCRIPTIONS_MAX_CONCURRENT_QUERIES)
```
Let's make this an instance attribute on `Executor` and add a `close` method to `Executor` that cleanly shuts down the executor. (I'm not sure this is totally necessary, but seems like a good idea.)

wedamija replied:
Do we want this as an instance variable though? At the moment, this will limit the total number of concurrent queries for the entire consumer, but if we make it per `Executor` then the limit only applies to each Executor instance... not sure whether we'll have more than one.

tkaemming replied:
I think it's reasonable to assume for now that there's only one `Executor` (note: maybe we should rename this to `SubscriptionExecutor` or `SubscriptionQueryExecutor` to avoid confusing with `ThreadPoolExecutor`) since by convention we run one consumer per dataset.

If we need to make sure that the total number of query execution threads started by the Python process are bounded, i'd rather do that by explicitly passing a `ThreadPoolExecutor` to the constructor and initializing it in the CLI where the consumer, worker, etc. are initialized (same idea as what we were talking about yesterday with the subscription store being passed the Redis client instead of depending on the module global.)

tkaemming commented on snuba/subscriptions/executor.py:
```diff
@@ -0,0 +1,38 @@
+    """
+
+    def __init__(self, dataset: Dataset):
+        self.dataset = dataset
```
Generally we name mangle instance attributes in Snuba (at least for newer code) unless they're intended to be used as public API:

```suggestion
        self.__dataset = dataset
```

tkaemming commented on snuba/subscriptions/executor.py:
```diff
@@ -0,0 +1,38 @@
+    def execute(
+        self, task: ScheduledTask[Subscription], tick: Tick, timer: Timer
+    ) -> Future[ClickhouseQueryResult]:
+        request: Request = task.task.build_request(
```
This type annotation isn't necessary, it should be able to be inferred by mypy.

wedamija replied:
Yeah I assumed that was the case, but maybe this sort of thing is a good idea for readability? Not totally sure, I suppose the editor should be able to infer this too.

tkaemming replied:
Yeah, the editor should be able to tell you. No strong opinion on readability if you want to leave this as-is — if it was me, I probably wouldn't have even made it a local, since `build_request` sort of describes the return type anyway.

tkaemming commented on snuba/subscriptions/executor.py:
```diff
@@ -0,0 +1,38 @@
+    to make the query, and the result is then returned as a `Future`.
+    """
+
+    def __init__(self, dataset: Dataset):
```
```suggestion
    def __init__(self, dataset: Dataset) -> None:
```

wedamija replied:
Do we really need to specify the return value of constructors?

tkaemming replied:
~Sometimes mypy complains if they are not defined, I haven't tracked down when/why (not sure if that was a prior version thing, or if it only needs to be done in certain circumstances) — it seems easier just to specify it and avoid any questions.~

Ok, you can disregard this, unless you'd like to add it for consistency; https://mypy.readthedocs.io/en/latest/class_basics.html#annotating-init-methods

tkaemming commented on snuba/subscriptions/executor.py:
```diff
@@ -0,0 +1,38 @@
+            tick.timestamps.upper,
+            tick.offsets.upper,
+            timer,
+        )
```
This will throw immediately, rather than binding the result of the future to an error, and will likely be uncaught by the worker/consumer.

I think we'll likely want to propagate some (maybe not all) of the errors that would be raised here to the caller instead (via the stream), since the failures here can include things like malformed queries which are expected to happen and shouldn't stop the progress of the consumer (and probably should be relayed back to the end user that their query no longer is valid.)

Here as an example from the work in progress implementation of the worker that demonstrates how I expect this to work (see the comment): https://github.com/getsentry/snuba/blob/28bf739541195b59851330a29a9663a9f2936b54/snuba/subscriptions/worker.py#L36-L48

We will at some point need to figure out how to separate expected errors (mostly bad queries, other things may fall into this category) and unexpected errors (bugs, possibly some networking problems) that should stop the consumer, but I think we can punt on that until we are actually implementing the codec that translates the `Future[ClickhouseQueryResult]` into the payload that goes on the stream.

tkaemming reviewed (changes requested):
Initial thoughts inline — this generally looks good, but I think we need to spend some more time thinking about error handling.

wedamija committed:
Merge branch 'master' into danf/subscription_executor

wedamija committed:
Use private variables.
Move `executor_pool` to an instance variable
Make sure all exceptions are set on the future

tkaemming reviewed (commented):
A couple of `mypy` errors inline. This also doesn't look like it was formatted with `black`?

Other than those minor things, I think this is good to go.

wedamija committed:
Merge branch 'master' into danf/subscription_executor

wedamija committed:
Black formatting
Mypy fixes

wedamija committed:
Inject ThreadPoolExecutor

wedamija committed:
Merge branch 'master' into danf/subscription_executor

wedamija commented:
Should be good to go. Black decided to format a couple of extra files for me.

tkaemming commented on snuba/subscriptions/executor.py:
```diff
@@ -0,0 +1,42 @@
+                self.__dataset, tick.timestamps.upper, tick.offsets.upper, timer
+            )
+        except Exception as e:
+            future = Future[ClickhouseQueryResult]()
```
```suggestion
            future: Future[ClickhouseQueryResult] = Future()
```

This will cause the "Future not subscriptable" thing you were seeing earlier.

tkaemming reviewed (changes requested):
I think this is good other than the one thing noted inline.

I'm curious why your formatter is being more aggressive than mine and would like to understand why but if we can't figure it out I guess it's not worth blocking on since I can run `black --check` on your branch (after merging master) without issues.

wedamija committed:
Revert "Black formatting"

This reverts commit 1bb3ebafd940d952d2c1df759c6c773a657df793.

wedamija committed:
Merge branch 'master' into danf/subscription_executor

tkaemming commented on snuba/subscriptions/executor.py:
```diff
@@ -0,0 +1,42 @@
+            )
+        return future
+
+    def close(self):
```
I think you lost your type annotation here in the merge or something:

```suggestion
    def close(self) -> None:
```

tkaemming commented on snuba/subscriptions/executor.py:
```diff
@@ -0,0 +1,42 @@
+        return future
+
+    def close(self):
+        self.__executor_pool.shutdown()
```
Kind of silly that `shutdown` is a blocking operation and not a future itself. 🤷‍♂

wedamija replied:
You can make it non-blocking with `wait=False`. The python program itself still has to wait until all the pending futures are done regardless though.

tkaemming reviewed (approved):
This looks good (aside from the quick fix inline.) 👍

wedamija committed:
Re-annotate
